File: backend/DashyBoard.Domain/Models/Reminder.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace DashyBoard.Domain.Models
{
    public class Reminder
    {
        public Guid Id { get; private set; }
        public Guid UserId { get; private set; }

        public string Title { get; private set; } = string.Empty;
        public string? Note { get; private set; }

        public DateTime DueAtUtc { get; private set; }

        public bool IsCompleted { get; private set; }
        public DateTime CreatedAtUtc { get; private set; }
        public DateTime? CompletedAtUtc { get; private set; }

        private Reminder() { } 

        public Reminder(Guid userId, string title, DateTime dueAtUtc, string? note = null)
        {
            if (string.IsNullOrWhiteSpace(title))
                throw new ArgumentException("Title is required.", nameof(title));

            Id = Guid.NewGuid();
            UserId = userId;
            Title = title.Trim();
            Note = string.IsNullOrWhiteSpace(note) ? null : note.Trim();
            DueAtUtc = dueAtUtc;
            CreatedAtUtc = DateTime.UtcNow;
            IsCompleted = false;
        }

        public void Update(string title, DateTime dueAtUtc, string? note = null)
        {
            if (string.IsNullOrWhiteSpace(title))
                throw new ArgumentException("Title is required.", nameof(title));

            Title = title.Trim();
            Note = string.IsNullOrWhiteSpace(note) ? null : note.Trim();
            DueAtUtc = dueAtUtc;
        }

        public void MarkCompleted()
        {
            if (IsCompleted) return;

            IsCompleted = true;
            CompletedAtUtc = DateTime.UtcNow;
        }

        public void MarkUncompleted()
        {
            IsCompleted = false;
            CompletedAtUtc = null;
        }
    }
}

File: frontend/src/api/types/reminder.ts
export interface ReminderDto {
    id: string;
    userId: string;
    title: string;
    note?: string;
    dueAtUtc: string;
    isCompleted: boolean;
    createdAtUtc: string;
    completedAtUtc?: string;
}

export interface CreateReminderCommand {
    title: string;
    dueAtUtc: string;
    note?: string;
}

export interface UpdateReminderCommand {
    id: string;
    title: string;
    dueAtUtc: string;
    note?: string;
}

export interface CompleteReminderCommand {
    id: string;
}

export interface DeleteReminderCommand {
    id: string;
}

export interface UncompleteReminderCommand {
    id: string;
}
